### DISCUSSION TITLE
Requesting feedback on unconventional usage of Mobx

### DISCUSSION THREAD

**Discussion by alshdavid** [mobxjs/mobx#2988] [Q&A]
I have implemented Mobx in an unconventional way and wanted to collect your thoughts (as creators, contributors and users) on the usage.

tl;dr: Wrapped Mobx within a React hook to enable simplified "view models" 

```jsx
class AppViewModel {
  title = 'Hello World'
}

const App = () => {
  const vm = useViewModel(() => new AppViewModel())

  useEffect(() => {
    setTimeout(() => vm.title = 'Updated title!', 1000)
  })

  return <div>
    <h1>{vm.title}</h1>
  </div>
}
```
_The above code will show "Hello World" then automatically update the text when the `vm.title` property is modified_

The reasoning for this approach is to allow for a separation of logic from React. The ViewModel is a virtual representation of the component "state" and can be tested independently of the component without requiring access to the document to render.

This approach also allows for the ability to use the simple property injection method of dependency injection where upstream services can be reactive and observed using mobx.

_Example. we could have a counter service dedicated to holding our counter value and modifying its current value:_

```javascript
class CounterService {
  value = 0

  increment() {
    this.value++;
  }

  decrement() {
    this.value--;
  }
}
```

Our component viewmodel can accept that service as a dependency and changes to the value can be automatically detected, propagating upwards triggering a view rerender.
```javascript
class AppViewModel {
  counterService

  constructor(
    counterService: CounterService
  ) {
    this.counterService = counterService
  }
}

const App = () => {
  // Use React context as a dependency container
  const counterService = useCounterService()
  const vm = useViewModel(() => new AppViewModel(counterService))

  return <div>
    <h1>{vm.counterService.value}</h1>
    <button onClick={() => vm.counterService.increment()}>Increment</button>
    <button onClick={() => vm.counterService.decrement()}>Decrement</button>
  </div>
}
```
This approach ensures a clean separation between the state of the application and the view renderer. 

This has implications on testing, allowing developers to bootstrap large portions of their front end without a document or with minimal document mocks. 

In cases of migration, if a developer should want to migrate to a newer version of the framework or to another framework (like Vue or similar)- they could take their existing view models with them and only need to rewrite the presentational layer.

This also ensures minimal code changes are required to for engineers to use Mobx in a project, minimising the mental overhead and improving the approachability of a project for new engineers.

The implementation of the hook is as follows:
```javascript
import { makeAutoObservable, configure } from "mobx";
import { deepObserve } from "mobx-utils";
import { useEffect, useState } from "react";

configure({
  enforceActions: "never",
  computedRequiresReaction: true,
  reactionRequiresObservable: true,
  observableRequiresReaction: false,
  disableErrorBoundaries: false
});

export function useViewModel(initFunc) {
  const [container, setContainer] = useState(() => ({
    viewModel: initFunc()
  }));

  useEffect(() => {
    container.viewModel = makeAutoObservable(container.viewModel);

    const dispose = deepObserve(container.viewModel, () => {
      // The object must have changed so force a rerender
      setContainer({ viewModel: container.viewModel });
    });

    return () => dispose();
  }, []);

  return container.viewModel;
}
```

When initialising services; they need to be hit with a healthy serving of `makeAutoObservable` prior to being supplied to their consumers.

**Sandbox**
[Here's a codesandbox.io example of what I am experimenting with](https://codesandbox.io/s/mobx-view-model-example-j6r8l?file=/src/App.tsx), including a healthy dose of unit tests

I have noticed this requires overriding a bunch of behaviours, does that mean bad things?
What are your thoughts? Have I gone off the deep end?

**Comment by IhsanMujdeci** [comment]
It seems like an issue with the non "observer component" route is that a component will re-render if it has a viewModel and something in it is changing that isn't referenced by the ui.
You can check this out for yourself by downloading react dev tools on chrome and going to components in dev tools and selecting following settings. This will highlight any components that re-render. 

![image](https://user-images.githubusercontent.com/12363843/123210526-73d7b000-d505-11eb-8c74-ddf1fd05f105.png)

I included an arbitrary setIncrement on a value not referenced in the ui in a view model class and am incrementing every second and a re-render is happening.

This re-rendering doesn't happen with observer component.

**Comment by mweststrate** [comment]
This model kills all optimisations that MobX offers; the deepObserve will response to everything that changes anywhere, whether it is relevant for the UI or not. I'm quite sure you could achieve the same in a much simpler and efficienter way without mobx, if you trap either actions or properties, and let them re-render components whenever something changes. If you search the issue tracker, you will find some past discussion where people want to get rid of `observer` and instead use `useObservable`, detailing why such an approach will loose MobX's edge.
